zebra-consensus/src/checkpoint.rs

```diff
@@ -459,6 +459,31 @@ where
             }
         };
 
+        // Check for a valid Merkle root. To prevent malleability (CVE-2012-2459),
+        // we also need to check whether the transaction hashes are unique.
+
+        let transaction_hashes = block
+            .transactions
+            .iter()
+            .map(|tx| tx.hash())
+            .collect::<Vec<_>>();
+        let merkle_root = transaction_hashes.iter().cloned().collect();
+
+        if block.header.merkle_root != merkle_root {
+            tx.send(Err(VerifyCheckpointError::BadMerkleRoot {
+                expected: block.header.merkle_root,
+                actual: merkle_root,
+            }))
+            .expect("rx has not been dropped yet");
+            return rx;
+        }
+
+        if transaction_hashes.len() != transaction_hashes.iter().collect::<HashSet<_>>().len() {
```

Comment:
Maybe a little comment that the HashSet is doing the deduping here

Reply:
Added a comment.